### PR TITLE
config/v1: operatorhub API improvements

### PR DESCRIPTION
**Pull request by aravindhp**
- Improve godocs
- Replace name and configuration map with HubSource that will reflect the current state of the default OperatorSources

### CODE REVIEW

**Comment by aravindhp** [comment]
@shawn-hurley I have made the changes you requested. Please take a look.

File: config/v1/types_operatorhub.go [hunk 1]
```diff
@@ -6,7 +6,10 @@ import (
 
 // OperatorHubSpec defines the desired state of OperatorHub
 type OperatorHubSpec struct {
-	// hubSources is the list of default OperatorSources and their configuration
+	// hubSources is the list of default OperatorSources and their configuration.
+	// If the list is empty, it indicates that the default OperatorSources are
+	// enabled on the cluster. The list of default OperatorSources and their
+	// current disabled state will always be reflected in the status block.
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Done

File: config/v1/types_operatorhub.go [hunk 2]
```diff
@@ -58,8 +61,8 @@ type HubSourceStatus struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:Required
 	Name string `json:"name,omitempty"`
-	// configuration is the state of the default OperatorSources configuration
-	Configuration map[string]string `json:"configuration,omitempty"`
+	// defaultHubSources reflects the disabled state of the default OperatorSources
```
[the remark this answers is not in the document]
Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Done

File: config/v1/types_operatorhub.go [hunk 3]
```diff
@@ -6,19 +6,27 @@ import (
...
 
-// OperatorHubStatus defines the observed state of OperatorHub
+// OperatorHubStatus defines the observed state of OperatorHub. The current
+// state of the default hub sources will always be reflected here.
 type OperatorHubStatus struct {
-	// hubSourcesStatus encapsulates the result applying the configuration
-	HubSourcesStatus []HubSourceStatus `json:"hubSourcesStatus,omitempty"`
+	// Sources encapsulates the result of applying the configuration for each
+	// hub source
+	Sources []HubSourceStatus `json:"Sources,omitempty"`
```
Comment (shawn-hurley) [inline config/v1/types_operatorhub.go]: lower case json

Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Fixed.

File: config/v1/types_operatorhub.go [hunk 4]
```diff
@@ -38,30 +46,24 @@ type OperatorHubList struct {
...
-	// +kubebuilder:validation:MaxLength=253
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:Required
-	Name string `json:"name,omitempty"`
-	// configuration is the state of the default OperatorSources configuration
-	Configuration map[string]string `json:"configuration,omitempty"`
+	HubSource
 	// status indicates success or failure in applying the configuration
-	Status string `json:"status,omitempty"`
+	Status string `json:"status"`
```
Comment (derekwaynecarr) [inline config/v1/types_operatorhub.go]: is this required?

Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Yes, we always want to reflect current status of all the default hub sources.

File: config/v1/types_operatorhub.go [hunk 5]
```diff
@@ -6,19 +6,27 @@ import (
...
 // OperatorHubSpec defines the desired state of OperatorHub
 type OperatorHubSpec struct {
-	// hubSources is the list of default OperatorSources and their configuration
-	HubSources []HubSource `json:"hubSources,omitempty"`
+	// sources is the list of default hub sources and their configuration.
+	// If the list is empty, it indicates that the default hub sources are
+	// enabled on the cluster. The list of default hub sources and their
+	// current state will always be reflected in the status block.
+	// +optional
+	Sources []HubSource `json:"sources,omitempty"`
```
Comment (smarterclayton) [inline config/v1/types_operatorhub.go]: Can I disable all hub sources?

**Comment by smarterclayton** [comment]
This doesn't look like it allows someone to disable all hub sources?

File: config/v1/types_operatorhub.go (hunk 5 above)
Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Yes you can

**Comment by aravindhp** [comment]
>This doesn't look like it allows someone to disable all hub sources?

Why do you think so? You can put all three defaults in HubSources and mark them as disabled, then they will be removed from the cluster.

File: config/v1/types_operatorhub.go (hunk 4 above)
Comment (aravindhp) [reply config/v1/types_operatorhub.go]: Thinking more about this I realize this is not requited. I will fix this in a follow up PR.

**Comment by aravindhp** [comment]
@smarterclayton, here is what the `cluster` `OperatorHub` object look on a fresh cluster:
```
(default-config {2} U:4 ✗) operator-marketplace oc get operatorhub cluster -o yaml
apiVersion: config.openshift.io/v1
kind: OperatorHub
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"config.openshift.io/v1","kind":"OperatorHub","metadata":{"annotations":{"release.openshift.io/create-only":"true"},"name":"cluster"},"spec":{}}
    release.openshift.io/create-only: "true"
  creationTimestamp: "2019-08-10T13:35:54Z"
  generation: 3
  name: cluster
  resourceVersion: "368836"
  selfLink: /apis/config.openshift.io/v1/operatorhubs/cluster
  uid: c6b59db1-bb73-11e9-8d5e-0277548b887e
spec: {}
status:
  Sources:
  - disabled: false
    name: redhat-operators
    status: Success
  - disabled: false
    name: certified-operators
    status: Success
  - disabled: false
    name: community-operators
    status: Success
```
Now if we want to disable all three default sources, this is what it looks like:
```
(default-config {2} U:4 ✗) operator-marketplace oc get operatorhub cluster -o yaml
apiVersion: config.openshift.io/v1
kind: OperatorHub
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"config.openshift.io/v1","kind":"OperatorHub","metadata":{"annotations":{},"name":"cluster"},"spec":{"sources":[{"disabled":true,"name":"community-operators"},{"disabled":true,"name":"redhat-operators"},{"disabled":true,"name":"certified-operators"}]}}
  creationTimestamp: "2019-08-10T13:35:54Z"
  generation: 4
  name: cluster
  resourceVersion: "369270"
  selfLink: /apis/config.openshift.io/v1/operatorhubs/cluster
  uid: c6b59db1-bb73-11e9-8d5e-0277548b887e
spec:
  sources:
  - disabled: true
    name: community-operators
  - disabled: true
    name: redhat-operators
  - disabled: true
    name: certified-operators
status:
  Sources:
  - disabled: true
    name: redhat-operators
    status: Success
  - disabled: true
    name: certified-operators
    status: Success
  - disabled: true
    name: community-operators
    status: Success
```
And to prove the sources are not present on the cluster:
```
(default-config {2} U:4 ✗) operator-marketplace oc get operatorsources -n openshift-marketplace 
No resources found.
```
Please let me know if you have more questions.

File: config/v1/types_operatorhub.go [hunk 6]
```diff
@@ -38,30 +46,24 @@ type OperatorHubList struct {
...
 // configuration to a default source
 type HubSourceStatus struct {
-	// name is the name of one of the default OperatorSources
-	// +kubebuilder:validation:MaxLength=253
-	// +kubebuilder:validation:MinLength=1
-	// +kubebuilder:Required
-	Name string `json:"name,omitempty"`
-	// configuration is the state of the default OperatorSources configuration
-	Configuration map[string]string `json:"configuration,omitempty"`
+	HubSource
```
Comment (aravindhp) [inline config/v1/types_operatorhub.go]: Not having a json tag is an issue when generating `openAPIValidation` block in CCO. I will fix this and whatever other changes are requested in a follow up PR.

**Comment by smarterclayton** [comment]
How do I know what defaults to add to mark them all disabled?

The following scenario has to be possible:

1. A 4.2 user installs a cluster and disables all hub sources (using whatever mechanism is available)
2. In ANY FUTURE 4.Y version, if a new hub source is introduced, it has to be disabled, because the user's intent was "disable all hub sources"

If the mechanism you currently have doesn't allow that to happen, we need to change the shape of the API.

**Comment by aravindhp** [comment]
I see your point now. I will spin up a follow up PR that adds to the API that allows a user to do this.

**Comment by aravindhp** [comment]
@smarterclayton I have addressed your requirement in https://github.com/openshift/api/pull/409

**Pull request merged**